pkg/helm/watcher/controller/helm_watcher.go

```diff
@@ -0,0 +1,94 @@
+package controller
+
+import (
+	"context"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	"github.com/go-logr/logr"
+	"github.com/helm/helm/pkg/chartutil"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/weaveworks/weave-gitops/pkg/helm"
+	"github.com/weaveworks/weave-gitops/pkg/helm/watcher/cache"
+)
+
+// HelmWatcherReconciler runs the `reconcile` loop for the watcher.
+type HelmWatcherReconciler struct {
+	client.Client
+	Scheme      *runtime.Scheme
+	Cache       cache.Cache
+	RepoManager helm.HelmRepoManager
+}
+
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories,verbs=get;list;watch
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories/status,verbs=get
+
+// Reconcile is either called when there is a new HelmRepository or, when there is an update to a HelmRepository.
+// Because the watcher watches all helmrepositories, it will update data for all of them.
+func (r *HelmWatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+
+	// get source object
+	var repository sourcev1.HelmRepository
+	if err := r.Get(ctx, req.NamespacedName, &repository); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if repository.Status.Artifact == nil {
+		// this should not even occur, because create already checks this, but we do this nevertheless.
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("found the repository: ", "name", repository.Name)
+	// Reconcile is called for two reasons. One, the repository was just created, two there is a new revision.
+	// Because of that, we don't care what's in the cache. We will always fetch and set it.
+
+	charts, err := r.RepoManager.GetCharts(context.Background(), &repository, helm.Profiles)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	values := make(cache.ValueMap)
+
+	for _, chart := range charts {
+		for _, v := range chart.AvailableVersions {
+			// what happens when there are no values? We should just skip that version...
+			valueBytes, err := r.RepoManager.GetValuesFile(context.Background(), &repository, &helm.ChartReference{
+				Chart:   chart.Name,
+				Version: v,
+			}, chartutil.ValuesfileName)
+
+			if err != nil {
+				log.Error(err, "failed to get values for chart and version, skipping...", "chart", chart.Name, "version", v)
+				// log and skip version
+				continue
+			}
+
+			values[chart.Name] = map[string][]byte{
+				v: valueBytes,
+			}
+		}
+	}
+
+	data := cache.Data{
+		Profiles: charts,
+		Values:   values,
```

Comment:
might be easier to map charts & values together, rather than having them separated?

Reply:
I found this way to be stupid easy to handle. And returning the right values is simpler and faster without having to rummage through profiles. What easier handling would you get out of bringing them together and in what way would you do that? Also, they are separate because I'm writing out profiles by yaml serialising them.

Reply:
is it easier?

Its currentyl
```
		Profiles: []*pb.Profile{
                 &pb.Profile{
		  Name:              "test-profiles-1",
		  Home:              "home",
		  Description:       "description",
		  Icon:              "icon",
		  KubeVersion:       "1.21",
		  AvailableVersions: []string{"0.0.1"},
	        }, 
                 &pb.Profile{
		  Name:              "test-profiles-2",
		  Home:              "home",
		  Description:       "description",
		  Icon:              "icon",
		  KubeVersion:       "1.21",
		  AvailableVersions: []string{"0.0.4"},
	        }, 
		Values: map[string]map[string][]byte{
			profile1.Name: {
				"0.0.1": []byte("value"),
			},
			profile2.Name: {
				"0.0.4": []byte("value"),
			},
		},
```

Woudn't:
```
Profiles: []Profile{
               {
                   &pb.Profile{
		    Name:              "test-profiles-1",
		    Home:              "home",
		    Description:       "description",
		    Icon:              "icon",
		    KubeVersion:       "1.21",
		    AvailableVersions: []string{"0.0.1"},
	          }, 
                  Values:
		    "0.0.1": []byte("value"),
		  },
	       },
               {
                   &pb.Profile{
		    Name:              "test-profiles-1",
		    Home:              "home",
		    Description:       "description",
		    Icon:              "icon",
		    KubeVersion:       "1.21",
		    AvailableVersions: []string{"0.0.1"},
	          }, 
                  Values:
		    "0.0.1": []byte("value"),
		  },
	       },
	     }    
```

be an easier mental model?